pkg/client/query/paramsquerier.go

```diff
@@ -0,0 +1,148 @@
+
+var _ client.ParamsQuerier[cosmostypes.Msg] = (*abstractParamsQuerier)(nil)
+
+// paramsQuerierIface is an interface which generated query clients MUST implement
+// to be compatible with the cachedParamsQuerier.
+//
+// DEV_NOTE: It is mainly required due to syntactic constraints imposed by the generics
+// (i.e. otherwise, P here MUST be a value type, and there's no way to express that Q
+// (below) SHOULD be in terms of the concrete type of P in NewCachedParamsQuerier).
+type paramsQuerierIface[P cosmostypes.Msg] interface {
+	GetParams(context.Context) (P, error)
+}
+
+// NewCachedParamsQuerier creates a new, generic, params querier with the given
+// concrete query client constructor and the configuration which results from
+// applying the given options.
+func NewCachedParamsQuerier[P cosmostypes.Msg, Q paramsQuerierIface[P]](
+	deps depinject.Config,
+	queryClientConstructor func(conn gogogrpc.ClientConn) Q,
+	opts ...ParamsQuerierOptionFn,
+) (_ client.ParamsQuerier[P], err error) {
+	cfg := DefaultParamsQuerierConfig()
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	paramsCache, err := cache.NewInMemoryCache[P](cfg.CacheOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	querier := &cachedParamsQuerier[P, Q]{
+		config:      cfg,
+		paramsCache: paramsCache,
+	}
+
+	if err = depinject.Inject(
+		deps,
+		&querier.clientConn,
+	); err != nil {
+		return nil, err
+	}
+
+	querier.queryClient = queryClientConstructor(querier.clientConn)
+
+	return querier, nil
+}
+
+// cachedParamsQuerier provides a generic implementation of cached param querying.
+// It handles parameter caching and chain querying in a generic way, where
+// P is a pointer type of the parameters, and Q is the interface type of the
+// corresponding query client.
+type cachedParamsQuerier[P cosmostypes.Msg, Q paramsQuerierIface[P]] struct {
+	clientConn  gogogrpc.ClientConn
+	queryClient Q
+	paramsCache client.HistoricalQueryCache[P]
+	config      *paramsQuerierConfig
+}
+
+// GetParams returns the latest cached params, if any; otherwise, it queries the
+// current on-chain params and caches them.
+func (bq *cachedParamsQuerier[P, Q]) GetParams(ctx context.Context) (P, error) {
+	logger := polylog.Ctx(ctx).With(
+		"module", bq.config.ModuleName,
+		"method", "GetParams",
+	)
+
+	// Check the cache first.
+	var paramsZero P
+	cached, err := bq.paramsCache.Get("params")
+	switch {
+	case err == nil:
+		logger.Debug().Msgf("params cache hit")
+		return cached, nil
+	case !errors.Is(err, cache.ErrCacheMiss):
+		return paramsZero, err
+	}
+
+	logger.Debug().Msgf("%s", err)
+
+	// Query on-chain on cache miss.
+	params, err := bq.queryClient.GetParams(ctx)
+	if err != nil {
+		if bq.config.ModuleParamError != nil {
+			return paramsZero, bq.config.ModuleParamError.Wrap(err.Error())
+		}
+		return paramsZero, err
+	}
+
+	// Update the cache.
+	if err = bq.paramsCache.Set("params", params); err != nil {
```

Comment:
We should consider the case where `paramsCache` is in `historical` mode, otherwise this call might fail.

Additionally, I think we should find a way to get the height/version of the retrieved params in order to set it using `SetAsOfVersion`.

Maybe add the `height` to the `QueryParamsResponse`?

Reply:
As discussed:

1. `cachedParamsQuerier`'s cache is ALWAYS in historical mode.
2. This isn't possible anymore due to the decision we made to disable `#Set()` in historical mode.
3. We're going to transition to event-driven cache warming.

Reply:
> We're going to transition to event-driven cache warming.

Can you add a TODO for that somewhere?

Reply:
> cachedParamsQuerier's cache is ALWAYS in historical mode.

#PUC